### PR TITLE
[AIRFLOW-2310]: Add AWS Glue Job Compatibility to Airflow

### PR DESCRIPTION
**Pull request by OElesin**
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/projects/AIRFLOW/issues/AIRFLOW-2310
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
Currently, there is no integration with AWS Glue Jobs. My pull request is basically an improvement to integrate running AWS Glue jobs with Airflow.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Added tests.contrib.test_aws_glue_job_hook.py.
However, moto the test class for boto3 does not support AWS Glue mock currently

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by Fokko** [comment]
Thanks @OElesin 
Can you squash your commits into one single commit?

**Comment by OElesin** [comment]
@Fokko, kindly review. 

Thanks

**Comment by Fokko** [comment]
@OElesin something must have gone wrong with the rebasing/merging, I see a lot of unrelated changes.

**Comment by OElesin** [comment]
@Fokko, this was the same problem I emphasized with the previous PR. Can you please point out the unrelated changes?

**Comment by Fokko** [comment]
It isn't clean. If you look at the Files tab, you see a lot of files changed related to GCP. That doesn't make sense.

**Comment by suma-ps** [comment]
@OElesin  Do you plan to resolve the merge issues soon? Looking forward to using the Glue operator soon, thanks!

**Comment by cnidus** [comment]
@OElesin I'm also really interested in this submission.

**Comment by OElesin** [comment]
@suma-ps and @cnidus ,

I have plans of working fixing this. But I kinda get squashing commits mixed up at times. I would try to resolve this by weekend and resubmit the PR.

Thanks @all

File: airflow/contrib/aws_glue_job_hook.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            RunId=run_id,
+            PredecessorsIncluded=True
+        )
+        job_run_state = job_status['JobRun']['JobRunState']
+        failed = job_run_state == 'FAILED'
+        stopped = job_run_state == 'STOPPED'
+        completed = job_run_state == 'SUCCEEDED'
+
+        while True:
+            if failed or stopped or completed:
```
Comment (xoen) [inline airflow/contrib/aws_glue_job_hook.py:153]: Just "accidentally" got here while looking at something else but I thought it was worth commenting on this loop.

Isn't this a potentially infinite loop? Once in the loop, how are these 3 variables updated? I don't see the call to get the up-to-date status after the `time.sleep()`.

**Comment by sid88in** [comment]
@OElesin AWS Glue Operator will be a great addition! Let's fix this PR and get this out?

**Comment by simonvanderveldt** [comment]
@OElesin Have you been able to make any progress on this?
There's one duplicate file in there `airflow/contrib/aws_glue_job_hook.py` vs `airflow/contrib/hook/aws_glue_job_hook.py`.

[edit] From looking at the PR it seems it will both create a job (using boto3's `create_job()`) as well as run the job (using boto3's `start_job_run()`). IMHO at the very least these should be separate things so one can run jobs without creating them.
To expand on this, we create the jobs using a normal development workflow/CI and we just want to be able to run them from Airflow. IMHO it's not Airflow's responsibility to create resources, it should only invoke them.

FYI All the unrelated changes are because you merged `master` into your branch instead of rebasing your branch on top of `master`

**Comment by oelesinsc24** [comment]
> @OElesin Have you been able to make any progress on this?
> There's one duplicate file in there `airflow/contrib/aws_glue_job_hook.py` vs `airflow/contrib/hook/aws_glue_job_hook.py`.
> 
> [edit] From looking at the PR it seems it will both create a job (using boto3's `create_job()`) as well as run the job (using boto3's `start_job_run()`). IMHO at the very least these should be separate things so one can run jobs without creating them.
> To expand on this, we create the jobs using a normal development workflow/CI and we just want to be able to run them from Airflow. IMHO it's not Airflow's responsibility to create resources, it should only invoke them.
> 
> FYI All the unrelated changes are because you merged `master` into your branch instead of rebasing your branch on top of `master`

This is absolutely correct. And also makes the implementation way easier. I would make the necessary changes and add the commit.

File: airflow/contrib/aws_glue_job_hook.py (hunk 1 above)
Comment (oelesinsc24) [reply airflow/contrib/aws_glue_job_hook.py:153]: Well, it's not an infinite loop as it exits the loop once the job status is in any of these states:
`FAILED`, `STOPPED` or `SUCCEEDED`.

**Comment by oelesinsc24** [comment]
Everyone, I have moved this PR to: https://github.com/apache/incubator-airflow/pull/4068. I will close this later and ask that we track changes on the new PR.

Apologies for the inconveniences.

**Comment by Fokko** [comment]
Moved to #4068

File: airflow/contrib/aws_glue_job_hook.py (hunk 1 above)
Comment (xoen) [reply airflow/contrib/aws_glue_job_hook.py:153]: Sorry @oelesinsc24 just saw this comment. I'm probably missing something but I can't see where the `failed`, `stopped`, `completed` variables change value inside the `while` loop. The `else` branch only sleeps, doesn't this mean that the `if` condition will always be false? 🤷‍♂

**Pull request closed without merging**